storage-node/packages/runtime-api/index.js

```diff
@@ -84,225 +91,247 @@ class RuntimeApi {
     }
 
-    const finalizedPromise = newExternallyControlledPromise()
+    // Functions to be called when the submitted transaction is finalized. They are initialized
+    // after the transaction is submitted to the resolve and reject function of the final promise
+    // returned by signAndSend
+    // on extrinsic success
+    let onFinalizedSuccess
+    // on extrinsic failure
+    let onFinalizedFailed
 
-    await this.executeWithAccountLock(accountId, async () => {
-      // Try to get the next nonce to use
-      let nonce = this.nonces[accountId]
+    // Function assigned when transaction is successfully submitted. Invoking it ubsubscribes from
+    // listening to tx status updates.
+    let unsubscribe
 
-      let incrementNonce = () => {
-        // only increment once
-        incrementNonce = () => {
-          /* turn it into a no-op */
-        }
-        nonce = nonce.addn(1)
-        this.nonces[accountId] = nonce
+    let lastTxUpdateResult
+
+    const handleTxUpdates = (result) => {
+      const { events = [], status } = result
+      if (!result || !status) {
+        return
       }
 
-      // If the nonce isn't available, get it from chain.
-      if (!nonce) {
-        // current nonce
-        // TODO: possible race condition here found by the linter
-        // eslint-disable-next-line require-atomic-updates
-        nonce = await this.api.query.system.accountNonce(accountId)
-        debug(`Got nonce for ${accountId} from chain: ${nonce}`)
+      lastTxUpdateResult = result
+
+      // Deal with statuses which will prevent
+      // extrinsic from finalizing.
+      if (status.isUsurped) {
+        debug(status.type)
+        debug(JSON.stringify(status.asUsurped))
+        onFinalizedFailed && onFinalizedFailed({ err: 'Usurped', result, tx: status.asFinalized })
       }
 
-      return new Promise((resolve, reject) => {
-        debug('Signing and sending tx')
-        // send(statusUpdates) returns a function for unsubscribing from status updates
-        const unsubscribe = tx
-          .sign(fromKey, { nonce })
-          .send(({ events = [], status }) => {
-            debug(`TX status: ${status.type}`)
-
-            // Whatever events we get, process them if there's someone interested.
-            // It is critical that this event handling doesn't prevent
-            try {
-              if (subscribed && callback) {
-                const matched = RuntimeApi.matchingEvents(subscribed, events)
-                debug('Matching events:', matched)
-                if (matched.length) {
-                  callback(matched)
-                }
-              }
-            } catch (err) {
-              debug(`Error handling events ${err.stack}`)
-            }
+      if (status.isDropped) {
+        debug(status.type)
+        debug(JSON.stringify(status.asDropped))
+        onFinalizedFailed && onFinalizedFailed({ err: 'Dropped', result, tx: status.asFinalized })
```

Comment:
Can we use `asFinalized` with `isDropped` without causing an error?

Reply:
fixed in https://github.com/Joystream/joystream/pull/1063/commits/cd1bbb859081064b603dc7fdec7563e2d8dced3d and then in https://github.com/Joystream/joystream/pull/1063/commits/a779bc51f9b2ff2d18d7381471891367e5e60870